mdb_dump.1
.\" Copyright 2015-2017 Leonid Yuriev <leo@yuriev.ru>.
.\" Copyright 2014-2017 Howard Chu, Symas Corp. All Rights Reserved.
.\" Copyright 2015,2016 Peter-Service R&D LLC <http://billing.ru/>.
.\" Copying restrictions apply.  See COPYRIGHT/LICENSE.
.TH MDB_DUMP 1 "2014/06/20" "LMDB 0.9.14"
.SH NAME
mdb_dump \- LMDB environment export tool
.SH SYNOPSIS
.B mdb_dump
[\c
.BR \-V ]
[\c
.BI \-f \ file\fR]
[\c
.BR \-l ]
[\c
.BR \-n ]
[\c
.BR \-p ]
[\c
.BR \-a \ |
.BI \-s \ subdb\fR]
.BR \ envpath
.SH DESCRIPTION
The
.B mdb_dump
utility reads a database and writes its contents to the
standard output using a portable flat-text format
understood by the
.BR mdb_load (1)
utility.
.SH OPTIONS
.TP
.BR \-V
Write the library version number to the standard output, and exit.
.TP
.BR \-f \ file
Write to the specified file instead of to the standard output.
.TP
.BR \-l
List the databases stored in the environment. Just the
names will be listed, no data will be output.
.TP
.BR \-n
Dump an LMDB database which does not use subdirectories.
.TP
.BR \-p
If characters in either the key or data items are printing characters (as
defined by isprint(3)), output them directly. This option permits users to
use standard text editors and tools to modify the contents of databases.

Note: different systems may have different notions about what characters
are considered printing characters, and databases dumped in this manner may
be less portable to external systems.
.TP
.BR \-a
Dump all of the subdatabases in the environment.
.TP
.BR \-s \ subdb
Dump a specific subdatabase. If no database is specified, only the main database is dumped.
.SH DIAGNOSTICS
Exit status is zero if no errors occur.
Errors result in a non-zero exit status and
a diagnostic message being written to standard error.

Dumping and reloading databases that use user-defined comparison functions
will result in new databases that use the default comparison functions.
\fBIn this case it is quite likely that the reloaded database will be
damaged beyond repair permitting neither record storage nor retrieval.\fP

The only available workaround is to modify the source for the
.BR mdb_load (1)
utility to load the database using the correct comparison functions.
.SH "SEE ALSO"
.BR mdb_load (1)
.SH AUTHOR
Howard Chu of Symas Corporation <http://www.symas.com>


CHANGES
MDBX
	Add MDB_PREV_MULTIPLE
	Add error MDB_PROBLEM, replace some MDB_CORRUPTED
	Workarounds for glibc bugs: #21031 and 21032.

LMDB 0.9.20 Release Engineering
	Fix mdb_load with escaped plaintext (ITS#8558)
	Fix mdb_cursor_last / mdb_put interaction (ITS#8557)

LMDB 0.9.19 Release (2016/12/28)
	Fix mdb_env_cwalk cursor init (ITS#8424)
	Fix robust mutexes on Solaris 10/11 (ITS#8339)
	Fix MDB_GET_BOTH on non-dup record (ITS#8393)
	Optimize mdb_drop
	Fix xcursors after mdb_cursor_del (ITS#8406)
	Fix MDB_NEXT_DUP after mdb_cursor_del (ITS#8412)
	Fix mdb_cursor_put resetting C_EOF (ITS#8489)
	Fix mdb_env_copyfd2 to return EPIPE on SIGPIPE (ITS#8504)
	Fix mdb_env_copy with empty DB (ITS#8209)
	Fix behaviors with fork (ITS#8505)
	Fix mdb_dbi_open with mainDB cursors (ITS#8542)
	Fix F_NOCACHE on MacOS, error is non-fatal (ITS#7682)
	Documentation
		Cleanup doxygen nits
		Note reserved vs actual mem/disk usage


LMDB 0.9.18 Release (2016/02/05)
	already done for mdbx - Fix robust mutex detection on glibc 2.10-11 (ITS#8330)
	Fix page_search_root assert on FreeDB (ITS#8336)
	Fix MDB_APPENDDUP vs. rewrite(single item) (ITS#8334)
	n/a for mdbx - Fix mdb_copy of large files on Windows
	Fix subcursor move after delete (ITS#8355)
	Fix mdb_midl_shrink off-by-one (ITS#8363)
	n/a for mdbx - Check for utf8_to_utf16 failures (ITS#7992)
	Catch strdup failure in mdb_dbi_open
	Build
		already done for mdbx - Additional makefile var tweaks (ITS#8169)
	Documentation
		Add Getting Started page
		Update WRITEMAP description

LMDB 0.9.17 Release (2015/11/30)
	Fix ITS#7377 catch calloc failure
	Fix ITS#8237 regression from ITS#7589
	Fix ITS#8238 page_split for DUPFIXED pages
	Fix ITS#8221 MDB_PAGE_FULL on delete/rebalance
	Fix ITS#8258 rebalance/split assert
	Fix ITS#8263 cursor_put cursor tracking
	Fix ITS#8264 cursor_del cursor tracking
	Fix ITS#8310 cursor_del cursor tracking
	Fix ITS#8299 mdb_del cursor tracking
	Fix ITS#8300 mdb_del cursor tracking
	Fix ITS#8304 mdb_del cursor tracking
	Fix ITS#7771 fakepage cursor tracking
	Fix ITS#7789 ensure mapsize >= pages in use
	Fix ITS#7971 mdb_txn_renew0() new reader slots
	already done for mdbx - Fix ITS#7969 use __sync_synchronize on non-x86
	Fix ITS#8311 page_split from update_key
	Fix ITS#8312 loose pages in nested txn
	Fix ITS#8313 mdb_rebalance dummy cursor
	Fix ITS#8315 dirty_room in nested txn
	Fix ITS#8323 dirty_list in nested txn
	Fix ITS#8316 page_merge cursor tracking
	Fix ITS#8319 mdb_load error messages
	Fix ITS#8320 mdb_load plaintext input
	Fix ITS#8321 cursor tracking
	Added mdb_txn_id() (ITS#7994)
	Added robust mutex support
	Miscellaneous cleanup/simplification
	Build
		Create install dirs if needed (ITS#8256)
		not affected mdbx - Fix ThreadProc decl on Win32/MSVC (ITS#8270)
		not affected mdbx - Added ssize_t typedef for MSVC (ITS#8067)
		not affected mdbx - Use ANSI apis on Windows (ITS#8069)
		already done for mdbx - Use O_SYNC if O_DSYNC,MDB_DSYNC are not defined (ITS#7209)
		already done for mdbx - Allow passing AR to make (ITS#8168)
		Allow passing mandir to make install (ITS#8169)

LMDB 0.9.16 Release (2015/08/14)
	Fix cursor EOF bug (ITS#8190)
	Fix handling of subDB records (ITS#8181)
	Fix mdb_midl_shrink() usage (ITS#8200)
	not affected mdbx - fix reference to EINTR on WIN32 from ITS#8106 (ITS#8192)

LMDB 0.9.15 Release (2015/06/19)
	Fix txn init (ITS#7961,#7987)
	Fix MDB_PREV_DUP (ITS#7955,#7671)
	Fix compact of empty env (ITS#7956)
	Fix mdb_copy file mode
	Fix mdb_env_close() after failed mdb_env_open()
	Fix mdb_rebalance collapsing root (ITS#8062)
	Fix mdb_load with large values (ITS#8066)
	Fix to retry writes on EINTR (ITS#8106)
	Fix mdb_cursor_del on empty DB (ITS#8109)
	Fix and Rework comparison for MDB_INTEGERKEY/MDB_INTEGERDUP (ITS#8117)
	Fix error handling (ITS#7959,#8157,etc.)
	Fix race conditions (ITS#7969,7970)
	Added workaround for fdatasync bug in ext3fs
	Build
		Don't use -fPIC for static lib
		Update .gitignore (ITS#7952,#7953)
		Cleanup for "make test" (ITS#7841), "make clean", mtest*.c
		Misc. Android/Windows cleanup
	Documentation
		Fix MDB_APPEND doc
		Fix MDB_MAXKEYSIZE doc (ITS#8156)
		Fix mdb_cursor_put,mdb_cursor_del EACCES description
		Fix mdb_env_sync(MDB_RDONLY env) doc (ITS#8021)
		Clarify MDB_WRITEMAP doc (ITS#8021)
		Clarify mdb_env_open doc
		Clarify mdb_dbi_open doc

LMDB 0.9.14 Release (2014/09/20)
	Fix to support 64K page size (ITS#7713)
	Fix to persist decreased as well as increased mapsizes (ITS#7789)
	Fix cursor bug when deleting last node of a DUPSORT key
	Fix mdb_env_info to return FIXEDMAP address
	Fix ambiguous error code from writing to closed DBI (ITS#7825)
	Fix mdb_copy copying past end of file (ITS#7886)
	Fix cursor bugs from page_merge/rebalance
	Fix to dirty fewer pages in deletes (mdb_page_loose())
	Fix mdb_dbi_open creating subDBs (ITS#7917)
	Fix mdb_cursor_get(_DUP) with single value (ITS#7913)
	Fix Windows compat issues in mtests (ITS#7879)
	Add compacting variant of mdb_copy
	Add BigEndian integer key compare code
	Add mdb_dump/mdb_load utilities

LMDB 0.9.13 Release (2014/06/18)
	Fix mdb_page_alloc unlimited overflow page search
	Documentation
		Re-fix MDB_CURRENT doc (ITS#7793)
		Fix MDB_GET_MULTIPLE/MDB_NEXT_MULTIPLE doc

LMDB 0.9.12 Release (2014/06/13)
	Fix MDB_GET_BOTH regression (ITS#7875,#7681)
	Fix MDB_MULTIPLE writing multiple keys (ITS#7834)
	Fix mdb_rebalance (ITS#7829)
	Fix mdb_page_split (ITS#7815)
	Fix md_entries count (ITS#7861,#7828,#7793)
	Fix MDB_CURRENT (ITS#7793)
	Fix possible crash on Windows DLL detach
	Misc code cleanup
	Documentation
		mdb_cursor_put: cursor moves on error (ITS#7771)


LMDB 0.9.11 Release (2014/01/15)
	Add mdb_env_set_assert() (ITS#7775)
	Fix: invalidate txn on page allocation errors (ITS#7377)
	Fix xcursor tracking in mdb_cursor_del0() (ITS#7771)
	Fix corruption from deletes (ITS#7756)
	Fix Windows/MSVC build issues
	Raise safe limit of max MDB_MAXKEYSIZE
	Misc code cleanup
	Documentation

mdb_dump.c
/* mdb_dump.c - memory-mapped database dump tool */

/*
 * Copyright 2015-2017 Leonid Yuriev <leo@yuriev.ru>.
 * Copyright 2011-2017 Howard Chu, Symas Corp.
 * Copyright 2015,2016 Peter-Service R&D LLC.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted only as authorized by the OpenLDAP
 * Public License.
 *
 * A copy of this license is available in the file LICENSE in the
 * top-level directory of the distribution or, alternatively, at
 * <http://www.OpenLDAP.org/license.html>.
 */

#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <unistd.h>
#include <signal.h>
#include "mdbx.h"

#define PRINT	1
static int mode;

typedef struct flagbit {
	int bit;
	char *name;
} flagbit;

flagbit dbflags[] = {
	{ MDB_REVERSEKEY, "reversekey" },
	{ MDB_DUPSORT, "dupsort" },
	{ MDB_INTEGERKEY, "integerkey" },
	{ MDB_DUPFIXED, "dupfixed" },
	{ MDB_INTEGERDUP, "integerdup" },
	{ MDB_REVERSEDUP, "reversedup" },
	{ 0, NULL }
};

static volatile sig_atomic_t gotsig;

static void dumpsig( int sig )
{
	(void) sig;
	gotsig = 1;
}

static const char hexc[] = "0123456789abcdef";

static void hex(unsigned char c)
{
	putchar(hexc[c >> 4]);
	putchar(hexc[c & 0xf]);
}

static void text(MDB_val *v)
{
	unsigned char *c, *end;

	putchar(' ');
	c = v->mv_data;
	end = c + v->mv_size;
	while (c < end) {
		if (isprint(*c)) {
			putchar(*c);
		} else {
			putchar('\\');
			hex(*c);
		}
		c++;
	}
	putchar('\n');
}

static void byte(MDB_val *v)
{
	unsigned char *c, *end;

	putchar(' ');
	c = v->mv_data;
	end = c + v->mv_size;
	while (c < end) {
		hex(*c++);
	}
	putchar('\n');
}

/* Dump in BDB-compatible format */
static int dumpit(MDB_txn *txn, MDB_dbi dbi, char *name)
{
	MDB_cursor *mc;
	MDB_stat ms;
	MDB_val key, data;
	MDB_envinfo info;
	unsigned int flags;
	int rc, i;

	rc = mdb_dbi_flags(txn, dbi, &flags);
	if (rc) return rc;

	rc = mdb_stat(txn, dbi, &ms);
	if (rc) return rc;

	rc = mdb_env_info(mdb_txn_env(txn), &info);
	if (rc) return rc;

	printf("VERSION=3\n");
	printf("format=%s\n", mode & PRINT ? "print" : "bytevalue");
	if (name)
		printf("database=%s\n", name);
	printf("type=btree\n");
	printf("mapsize=%zu\n", info.me_mapsize);
	if (info.me_mapaddr)
		printf("mapaddr=%p\n", info.me_mapaddr);
	printf("maxreaders=%u\n", info.me_maxreaders);

	for (i=0; dbflags[i].bit; i++)
		if (flags & dbflags[i].bit)
			printf("%s=1\n", dbflags[i].name);

	printf("db_pagesize=%d\n", ms.ms_psize);
	printf("HEADER=END\n");

	rc = mdb_cursor_open(txn, dbi, &mc);
	if (rc) return rc;

	while ((rc = mdb_cursor_get(mc, &key, &data, MDB_NEXT)) == MDB_SUCCESS) {
		if (gotsig) {
			rc = EINTR;
			break;
		}
		if (mode & PRINT) {
			text(&key);
			text(&data);
		} else {
			byte(&key);
			byte(&data);
		}
	}
	printf("DATA=END\n");
	if (rc == MDB_NOTFOUND)
		rc = MDB_SUCCESS;

	return rc;
}

static void usage(char *prog)
{
	fprintf(stderr, "usage: %s [-V] [-f output] [-l] [-n] [-p] [-a|-s subdb] dbpath\n", prog);
	exit(EXIT_FAILURE);
}

int main(int argc, char *argv[])
{
	int i, rc;
	MDB_env *env;
	MDB_txn *txn;
	MDB_dbi dbi;
	char *prog = argv[0];
	char *envname;
	char *subname = NULL;
	int alldbs = 0, envflags = 0, list = 0;

	if (argc < 2) {
		usage(prog);
	}

	/* -a: dump main DB and all subDBs
	 * -s: dump only the named subDB
	 * -n: use NOSUBDIR flag on env_open
	 * -p: use printable characters
	 * -f: write to file instead of stdout
	 * -V: print version and exit
	 * (default) dump only the main DB
	 */
	while ((i = getopt(argc, argv, "af:lnps:V")) != EOF) {
		switch(i) {
		case 'V':
			printf("%s\n", MDB_VERSION_STRING);
			exit(0);
			break;
		case 'l':
			list = 1;
			/*FALLTHROUGH*/;
		case 'a':
			if (subname)
				usage(prog);
			alldbs++;
			break;
		case 'f':
			if (freopen(optarg, "w", stdout) == NULL) {
				fprintf(stderr, "%s: %s: reopen: %s\n",
					prog, optarg, strerror(errno));
				exit(EXIT_FAILURE);
			}
			break;
		case 'n':
			envflags |= MDB_NOSUBDIR;
			break;
		case 'p':
			mode |= PRINT;
			break;
		case 's':
			if (alldbs)
				usage(prog);
			subname = optarg;
			break;
		default:
			usage(prog);
		}
	}

	if (optind != argc - 1)
		usage(prog);

#ifdef SIGPIPE
	signal(SIGPIPE, dumpsig);
#endif
#ifdef SIGHUP
	signal(SIGHUP, dumpsig);
#endif
	signal(SIGINT, dumpsig);
	signal(SIGTERM, dumpsig);

	envname = argv[optind];
	rc = mdb_env_create(&env);
	if (rc) {
		fprintf(stderr, "mdb_env_create failed, error %d %s\n", rc, mdb_strerror(rc));
		return EXIT_FAILURE;
	}

	if (alldbs || subname) {
		mdb_env_set_maxdbs(env, 2);
	}

	rc = mdb_env_open(env, envname, envflags | MDB_RDONLY, 0664);
	if (rc) {
		fprintf(stderr, "mdb_env_open failed, error %d %s\n", rc, mdb_strerror(rc));
		goto env_close;
	}

	rc = mdb_txn_begin(env, NULL, MDB_RDONLY, &txn);
	if (rc) {
		fprintf(stderr, "mdb_txn_begin failed, error %d %s\n", rc, mdb_strerror(rc));
		goto env_close;
	}

	rc = mdb_open(txn, subname, 0, &dbi);
	if (rc) {
		fprintf(stderr, "mdb_open failed, error %d %s\n", rc, mdb_strerror(rc));
		goto txn_abort;
	}

	if (alldbs) {
		MDB_cursor *cursor;
		MDB_val key;
		int count = 0;

		rc = mdb_cursor_open(txn, dbi, &cursor);
		if (rc) {
			fprintf(stderr, "mdb_cursor_open failed, error %d %s\n", rc, mdb_strerror(rc));
			goto txn_abort;
		}
		while ((rc = mdb_cursor_get(cursor, &key, NULL, MDB_NEXT_NODUP)) == 0) {
			char *str;
			MDB_dbi db2;
			if (memchr(key.mv_data, '\0', key.mv_size))
				continue;
			count++;
			str = malloc(key.mv_size+1);
			memcpy(str, key.mv_data, key.mv_size);
			str[key.mv_size] = '\0';
			rc = mdb_open(txn, str, 0, &db2);
			if (rc == MDB_SUCCESS) {
				if (list) {
					printf("%s\n", str);
					list++;
				} else {
					rc = dumpit(txn, db2, str);
					if (rc)
						break;
				}
				mdb_close(env, db2);
			}
			free(str);
			if (rc) continue;
		}
		mdb_cursor_close(cursor);
		if (!count) {
			fprintf(stderr, "%s: %s does not contain multiple databases\n", prog, envname);
			rc = MDB_NOTFOUND;
		} else if (rc == MDB_INCOMPATIBLE) {
			/* LY: the record it not a named sub-db. */
			rc = MDB_SUCCESS;
		}
	} else {
		rc = dumpit(txn, dbi, subname);
	}
	if (rc && rc != MDB_NOTFOUND)
		fprintf(stderr, "%s: %s: %s\n", prog, envname, mdb_strerror(rc));

	mdb_close(env, dbi);
txn_abort:
	mdb_txn_abort(txn);
env_close:
	mdb_env_close(env);

	return rc ? EXIT_FAILURE : EXIT_SUCCESS;
}
